apps/extension/src/core/domains/balances/handler.ts

```diff
@@ -89,107 +89,101 @@ const subscribeBalancesByParams = async (
+
+        const staleBalances = Object.values(initBalanceMap)
+          .filter((b) => b.status === "initializing")
+          .map((b) => ({ ...b, status: "stale" } as BalanceJson))
+        callback({ type: "upsert", balances: new Balances(staleBalances).toJSON() })
+      }, 30_000)
+
+      // subscribe to balances by params
+      balancesUnsubCallbacks = balanceModules.map((balanceModule) =>
+        balanceModule.subscribeBalances(
+          addressesByTokenByModule[balanceModule.type] ?? {},
+          (error, result) => {
+            if (thisGeneration !== balancesSubscriptionsGeneration) return
+
+            // eslint-disable-next-line no-console
+            if (error) return DEBUG && console.error(error)
 
-  // restart subscriptions each type params update
-  obsSubscriptionParams.subscribe(async ({ addressesByTokenByModule }) => {
-    // close previous subscriptions
-    await Promise.all(closeSubscriptionCallbacks)
-
-    // create placeholder rows for all missing balances, so FE knows they are initializing
-    const initBalances: BalanceJson[] = []
-    for (const balanceModule of balanceModules) {
-      const addressesByToken = addressesByTokenByModule[balanceModule.type] ?? {}
-      for (const [tokenId, addresses] of Object.entries(addressesByToken))
-        for (const address of addresses)
-          initBalances.push(balanceModule.getPlaceholderBalance(tokenId, address))
-    }
-    const getBalanceKey = (b: BalanceJson | Balance) => `${b.tokenId}:${b.address}`
-    const initBalanceMap = Object.fromEntries(initBalances.map((b) => [getBalanceKey(b), b]))
-    callback({ type: "upsert", balances: new Balances(initBalances).toJSON() })
-
-    // after 30 seconds, change the status of all balances still initializing to stale
-    setTimeout(() => {
-      const staleBalances = Object.values(initBalanceMap)
-        .filter((b) => b.status === "initializing")
-        .map((b) => ({ ...b, status: "stale" } as BalanceJson))
-      callback({ type: "upsert", balances: new Balances(staleBalances).toJSON() })
-    }, 30_000)
-
-    // subscribe to balances by params
-    closeSubscriptionCallbacks = balanceModules.map((balanceModule) =>
-      balanceModule.subscribeBalances(
-        addressesByTokenByModule[balanceModule.type] ?? {},
-        (error, result) => {
-          // eslint-disable-next-line no-console
-          if (error) DEBUG && console.error(error)
-          else {
             for (const balance of result?.each ?? []) delete initBalanceMap[getBalanceKey(balance)]
             callback({ type: "upsert", balances: (result ?? new Balances([])).toJSON() })
           }
-        }
+        )
       )
-    )
+    },
   })
 
   // unsub on port disconnect
   port.onDisconnect.addListener((): void => {
     unsubscribe(id)
-    closeSubscriptionCallbacks.forEach((cb) => cb.then((close) => close()))
+    byParamsSubscription.unsubscribe()
+    balancesUnsubCallbacks.forEach((cb) => cb.then((close) => close()))
```

Comment:
Unsub all subscriptions (byParams as well as the inner balances subs) when the UI requesting these balances is closed